Keep the stack items as an instance of bn.js

axic opened the pull request.

axic commented:
Down to:

```
# tests 2412
# pass  2323
# fail  89
```

axic commented:
Rebased this locally over #174, will push once that is merged.

axic commented:
@jwasinger I think this should work now too. Rebased and fixed some bugs.

axic commented:
Fixed returndatasize, returndatacopy, jumpi and log. Push may be broken.

jwasinger commented:
`SSTORE` is also probably broken here:

https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/opFns.js#L401-L410

`{0}.length != {}.length` even though they should effectively be treated the same.  I tried to fix this but somehow the arguments to `SSTORE` were not BN (maybe this is because previous pushes (broken?) are not putting BNs on the stack somehow..)

axic commented:
@jwasinger I think I've added back the lost changes. Fixed DUP also to duplicate the stack item (and not just put a second reference there). It may work?

axic commented:
@holgerd77 do you want to take a look at this? This is the only remaining big internal cleanup PR.

holgerd77 commented:
@axic Yes, can do.

holgerd77 commented:
Current status:

```
1..4646
# tests 4646
# pass  3993
# fail  653
```

(Where do these additional 2000+ tests come from?)

holgerd77 commented:
Ok, that was a tiny change, bringing down the failure rate to a level where it was before above:

```
1..4646
# tests 4646
# pass  4551
# fail  95
```

Will now look into the remaining tests failing.

jwasinger commented:
> (Where do these additional 2000+ tests come from?)

Probably test cases for precompiles

holgerd77 commented:
Two other fixes, this is now down to 0 failures:

```
1..4646
# tests 4646
# pass  4646

# ok
```

😄 

Let's wait with merging at least until Monday though, maybe someone wants to have another one-step-back view on this.

holgerd77 commented:
Short speed comparison (on local MacBook Air) running the state tests:

``feature/bn-as-stack`` branch: 4:04 min
``master`` branch: 4:08 min

Hmm, seems this was more for easier reading than for speed though.

axic commented on lib/opFns.js:
```diff
@@ -310,8 +310,8 @@ module.exports = {
-    // NOTE: we're using a 'trick' here to get the least significant byte
-    byte = byte.toArrayLike(Buffer, 'le', 1)
+    byte = byte.toArrayLike(Buffer, 'be', 32)
+    byte = byte.slice(byte.length - 1)
```
Can you extract a case from the tests where this caused an issue? I'd like to figure out what is the actual reason.

holgerd77 replied:
``node tests/tester.js -s --test='randomStatetest237' --jsontrace``

In Node console:

```node
const utils = require('ethereumjs-util')
const BN = utils.BN
var bn = new BN('fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe', 'hex')
bn.toArrayLike(Buffer, 'le', 1)
```

gives

```
Error: byte array longer than desired length
    at assert (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/bn.js/lib/bn.js:6:21)
    at BN.toArrayLike (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/bn.js/lib/bn.js:527:5)
    at repl:1:4
    at ContextifyScript.Script.runInThisContext (vm.js:44:33)
    at REPLServer.defaultEval (repl.js:239:29)
    at bound (domain.js:301:14)
    at REPLServer.runBound [as eval] (domain.js:314:12)
    at REPLServer.onLine (repl.js:440:10)
    at emitOne (events.js:120:20)
    at REPLServer.emit (events.js:210:7)
```

holgerd77 replied:
Actually this is failing for every Buffer with length > 1, e.g. also for ``var bn = new BN('fffe', 'hex')``.

axic replied:
Ah true, never end the end added the truncation support to bn.js.

axic replied:
Could still do `bn.andl(0xff).toArrayLike(Buffer, 'le', 1)`

axic commented:
@holgerd77 that optimisation works - do you want to double check it?

holgerd77 commented:
Should we do a minor release after this is merged?

@axic: I have to admit I'm not getting your new trick. 😎 😸

axic commented:
`.andln` is bitwise AND, so basically we get the lowest 1 byte (with `0xff`) out of BN. But `.andln`, unlike any other bn.js function, doesn't return a new BN instance, rather a Javascript number.

Sure, we can make a release.

axic commented on lib/opFns.js:
```diff
@@ -122,22 +122,22 @@ module.exports = {
   },
   ISZERO: function (a, runState) {
-    return new BN(a.isZero())
+    return new BN(a.isZero() ? 1 : 0)
```
I think these make sense, but were there actually causing an issue?

holgerd77 replied:
Yes, creating a ``BN`` instance from ``true`` is actually resulting in another value initialization than with ``1``. 

```
> new BN(true)
<BN: 7d0e>
> new BN(1)
<BN: 1>
```

Not sure if this is an error in ``bn.js``. I find this more explicit and better to read anyway.

axic replied:
True, we are till using an outdated bn.js version which treated everything as a number and did random things. Finally a recent version would throw an error of invalid digit for given number base :)

axic commented on lib/opFns.js:
```diff
@@ -122,22 +122,22 @@ module.exports = {
   },
   EQ: function (a, b, runState) {
-    return new BN(a.cmp(b) === 0)
+    return new BN(a.cmp(b) === 0 ? 1 : 0)
```
Oh this code was done way before the higher level helpers were added to bn.js. We could use `a.eq(b)` here and above.

axic replied:
Created a new PR for this.

axic commented:
@holgerd77 are you OK with squashing down the `MSTORE8` changes and merging?

axic commented:
@holgerd77 I'll do the squashing of that one into the original, but leaving every other fix intact.

Just asking if you are happy with the last solution?

holgerd77 commented:
Yes. Just curious how you will do that (squashing into the original, skipping the in-between commits) Not seeing anything for that in interactive rebase command selection. Any trick?

axic commented:
Try interactive rebase: `git rebase -i HEAD~6` (offers interactive rebasing of the last 6 commits).

holgerd77 commented:
Ah, this note "These lines can be re-ordered;" from the rebase instructions was the missing part for me, didn't know that.

axic commented:
@holgerd77 ok to accept?

holgerd77 commented:
Yes (the "Yes" from above was also already answering this. Sorry maybe too hidden. :-)).

axic commented:
> Hmm, seems this was more for easier reading than for speed though.

I think there should be a more measurable speed increase on running contracts doing arithmetics. See some of @gcolvin's benchmarks for rc5, division and exponentiation.
